src/analysis/processing/qgsalgorithmrandompointsinpolygons.cpp

```diff
@@ -0,0 +1,394 @@
+    QgsSpatialIndex localIndex;
+    int localIndexPoints = 0;
+    int pointsAddedForThisFeature = 0;
+    // Get data defined parameters
+    int numberPointsForThisFeature = mNumPoints;
+    if ( mDynamicNumPoints )
+      numberPointsForThisFeature = mNumPointsProperty.valueAsInt( expressionContext, numberPointsForThisFeature );
+    desiredNumberOfPoints += numberPointsForThisFeature;
+    int maxAttemptsForThisFeature = mMaxAttempts;
+    if ( mDynamicMaxAttempts )
+      maxAttemptsForThisFeature = mMaxAttemptsProperty.valueAsInt( expressionContext, maxAttemptsForThisFeature );
+    double minDistanceForThisFeature = mMinDistance;
+    if ( mDynamicMinDistance )
+      minDistanceForThisFeature = mMinDistanceProperty.valueAsDouble( expressionContext, minDistanceForThisFeature );
+    const double pointProgressIncrement = featureProgressStep / ( numberPointsForThisFeature * maxAttemptsForThisFeature );
+    double pointProgress = 0.0;
+    unsigned long fseed = uniformIntDist( mt );
+    // Check if we can avoid using the acceptPoint function
+    if ( ( minDistanceForThisFeature == 0 ) && ( mMinDistanceGlobal == 0 ) )
+    {
+      QVector< QgsPointXY > newPoints = polyGeom.randomPointsInPolygon( numberPointsForThisFeature, fseed );
+      for ( int i = 0; i < newPoints.length(); i++ )
+      {
+        // add the point
+        QgsPointXY pt = newPoints[i];
+        QgsFeature f = QgsFeature( totNPoints );
+        QgsAttributes pAttrs = QgsAttributes();
+        pAttrs.append( totNPoints );
+        if ( mIncludePolygonAttr )
+        {
+          pAttrs.append( polyFeat.attributes() );
+        }
+        f.setAttributes( pAttrs );
+        QgsGeometry newGeom = QgsGeometry::fromPointXY( pt );
+        f.setGeometry( newGeom );
+        sink->addFeature( f, QgsFeatureSink::FastInsert );
+        totNPoints++;
+        pointsAddedForThisFeature++;
+        pointProgress += pointProgressIncrement * ( maxAttemptsForThisFeature );
+      }
+      feedback->setProgress( baseFeatureProgress + pointProgress );
+      continue;
+    }
+    else
+    {
+      // Have to check for minimum distance, provide the acceptPoints function
+      QVector< QgsPointXY > newPoints = polyGeom.randomPointsInPolygon( numberPointsForThisFeature, [ =, &globalIndex, &indexPoints ]( const QgsPointXY & newPoint ) mutable -> bool
+      {
+        // May have to check minimum distance to existing points
+        // The first point can always be added
+        // Global first
+        if ( mMinDistanceGlobal != 0.0 && indexPoints > 0 )
+        {
+          QList<QgsFeatureId> neighbors = globalIndex.nearestNeighbor( newPoint, 1, mMinDistanceGlobal );
+          if ( totNPoints > 0 && !neighbors.empty() )
+          {
+            return false;
+          }
+        }
+        // Then local (if larger than global)
+        if ( minDistanceForThisFeature != 0 && mMinDistanceGlobal < minDistanceForThisFeature && localIndexPoints > 0 )
```

Comment:
Possible optimisation opportunity: the local index is always going to be smaller than the global index, so checks against the local index will always be much faster than those made against the global index. If you moved this check to sit before the global index check you may squeeze out some more speed...